x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java

```diff
@@ -232,22 +219,32 @@ public List<Gap> waitForRange(final ByteRange range, final ByteRange subRange, f
             assert ranges.containsAll(pendingRanges) : ranges + " vs " + pendingRanges;
             assert pendingRanges.stream().allMatch(Range::isPending) : pendingRanges;
             assert pendingRanges.size() != 1 || gaps.size() <= 1 : gaps;
+        }
 
-            // Pending ranges that needs to be filled before executing the listener
-            requiredRanges = range.equals(subRange)
-                ? pendingRanges
-                : pendingRanges.stream()
-                    .filter(pendingRange -> pendingRange.start < subRange.end())
-                    .filter(pendingRange -> subRange.start() < pendingRange.end)
-                    .sorted(Comparator.comparingLong(r -> r.start))
-                    .collect(Collectors.toList());
+        // Pending ranges that needs to be filled before executing the listener
+        if (range.equals(subRange) == false) {
+            pendingRanges.removeIf(pendingRange -> (pendingRange.start < subRange.end() && subRange.start() < pendingRange.end) == false);
+            pendingRanges.sort(RANGE_START_COMPARATOR);
         }
 
-        subscribeToCompletionListeners(requiredRanges, subRange.end(), wrappedListener);
+        subscribeToCompletionListeners(pendingRanges, subRange.end(), wrappedListener);
 
         return Collections.unmodifiableList(gaps);
     }
 
+    private void determineStartingRange(ByteRange range, List<Range> pendingRanges, Range targetRange) {
```

Comment:
I was wondering if it would be more idiomatic to make `determineStartingRange` return `pendingRanges` instead of supplying a list container to the method. I believe it would be more clear what is mutated.

```java
List<Range> pendingRanges;
synchronized (mutex) {
      pendingRanges = determineStartingRange(range, targetRange);
      ...
}      
```

Reply:
Still good to instanciate the list outside of the synchronized block